CANTroller2/src/uictrl.h

```diff
@@ -1,9 +1,36 @@
 #ifndef UICTRL_H
 #define UICTRL_H
 
-#include "globals.h"
+#include "utils.h"
+#include <memory> // for std::shared_ptr
 #include "FunctionalInterrupt.h"
 
+// Potentiometer does an analog read from a pin and maps it to a percent (0%-100%). We filter the value to keep it smooth. The output can either be read as a value, or passed as a pointer so that other components can
+// access it directly as needed.
+class Potentiometer {
+    protected:
+        // static constexpr float adc_min = 300; // TUNED 230603 - Used only in determining theconversion factor
+        static constexpr float adc_min = 300; // TUNED 230603 - Used only in determining theconversion factor
+        static constexpr float adc_max = 4095; // TUNED 230613 - adc max measured = ?, or 9x.? % of adc_range. Used only in determining theconversion factor
+        static constexpr float _ema_alpha = 0.1;
+        uint8_t _pin;
+        std::shared_ptr<float> _val;
+    public:
+        static constexpr float min = 0.0; // percent min
+        static constexpr float max = 100.0; // percent max
+        Potentiometer(uint8_t arg_pin) : _pin(arg_pin), _val(std::make_shared<float>(0.0)) {}
+        Potentiometer() = delete; // must have a pin defined
+        void setup() {
+            set_pin(_pin, INPUT);
+        }
+        void update() {
+            float new_val = constrain(map(static_cast<float>(analogRead(_pin)), adc_min, adc_max, min, max), min, max);

```

Comment:
Nit just for readability, consider making a function, or using a class for these value conversions. eg
c```
lass AnalogInput {
  int _pin;
  int _value;

  float percent() {
    return map(_value, adcMin, adcMax, 0, 100); 
  }

```
or as a function name it something like mapAnalogInputToPercentage

Reply:
well this one is only used once here, but there are a lot of functions which map to pot input, so how about

float Potentiometer::mapToRange(Value min, Value max) {
    return map(*_val, 0.0, 100.0, min, max);
}

To be used like this:

tach_filt_rpm = pot.mapToRange(0.0, tach_govern_rpm);

Reply:
Yeah, I like that. I suppose if it's always from 0 to 100 for input min and input max, those could be private class variables that get set either with default values (eg all whatever devices have default _input_min = 0 an _input_max=100, and we could either set those to different values in the instantiation like mypot = Potentiometer(min=20,max=80) or using a setter method. Just to avoid the 0 and 100 not being named variables.

Reply:
yep i made them class variables called _percent_min and _percent_max. I don't think we need to add the ability to change those (until we actually need it).